### PR TITLE
Remove "pkg/controller/volume/scheduling" dependency from "pkg/scheduler/framework/plugins"

### PR DESCRIPTION
**Pull request by ikeeip**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This patch aims to simplify decoupling `pkg/scheduler/framework/plugins` package from internal `k8s.io/kubernetes packages`. 

#### Which issue(s) this PR fixes:

Refs: #89930, #102953

#### Special notes for your reviewer:

All dependencies of VolumeBinding plugin from `k8s.io/kubernetes/pkg/controller/volume/scheduling` package moved to
`k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding` package, and their test helpers was refactored:

- whole file pkg/controller/volume/scheduling/scheduler_assume_cache.go
- whole file pkg/controller/volume/scheduling/scheduler_assume_cache_test.go
- whole file pkg/controller/volume/scheduling/scheduler_binder.go
- whole file pkg/controller/volume/scheduling/scheduler_binder_fake.go
- whole file pkg/controller/volume/scheduling/scheduler_binder_test.go

Package `k8s.io/kubernetes/pkg/controller/volume/scheduling/metrics` moved to `k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding/metrics` because it only used in VolumeBinding plugin and (e2e) tests.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/volumebinding/OWNERS [hunk 1]
```diff
@@ -1,6 +1,11 @@
 # See the OWNERS docs at https://go.k8s.io/owners
 
+approvers:
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/OWNERS:3]: @Huang-Wei what are your thoughts on this?

Another option is to create folder within `plugins/volumebinding` which is completely owned by sig storage, but the parent folder is still owned by sig-scheduling. But perhaps that's overkill.

File: pkg/scheduler/framework/plugins/volumebinding/scheduler_assume_cache.go [hunk 2]
```diff
@@ -14,14 +14,15 @@ See the License for the specific language governing permissions and
...
-package scheduling
+package volumebinding
 
 import (
 	"fmt"
-	storagehelpers "k8s.io/component-helpers/storage/volume"
 	"strconv"
 	"sync"
 
+	storagehelpers "k8s.io/component-helpers/storage/volume"
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/scheduler_assume_cache.go:24]: put this in the import group below

File: pkg/scheduler/framework/plugins/volumebinding/test_utils.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+func pvWithPhase(phase v1.PersistentVolumePhase) pvOpt {
+	return func(pv *v1.PersistentVolume) error {
+		pv.Status = v1.PersistentVolumeStatus{
+			Phase: phase,
+		}
+		return nil
+	}
+}
+
+type pvcOpt func(pv *v1.PersistentVolumeClaim) error
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/volumebinding/test_utils.go:134]: Can we remove the returned error? as I don't spot any functions returning an error, also no error handling on the caller side.

File: pkg/scheduler/metrics/metrics.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ import (
 
 	"k8s.io/component-base/metrics"
 	"k8s.io/component-base/metrics/legacyregistry"
-	volumeschedulingmetrics "k8s.io/kubernetes/pkg/controller/volume/scheduling/metrics"
+	volumeschedulingmetrics "k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding/metrics"
```
Comment (Huang-Wei) [inline pkg/scheduler/metrics/metrics.go:25]: nit: `volumebindingmetrics` as the alias.

**Review by Huang-Wei — COMMENTED** [review]
Could you please rename all `volumebinding/scheduler_xyz.go` to  `volumebinding/xyz.go` as the context is clear enough?

File: pkg/scheduler/framework/plugins/volumebinding/OWNERS (hunk 1 above)
Comment (ikeeip) [reply pkg/scheduler/framework/plugins/volumebinding/OWNERS:3]: It is still not clear for me. We need remove this file to return ownership to sig-scheduling or we need to do something like this:
```
approvers:
- sig-scheduling-maintainers
reviewers:
- sig-scheduling
- cofyc
- msau42
- lichuqiang
```
(leave original reviewers 
https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/pkg/controller/volume/scheduling/OWNERS)

WDYT @Huang-Wei ?

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/volumebinding/OWNERS:3]: You can leave it like this

File: pkg/scheduler/framework/plugins/volumebinding/test_utils.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
+	"k8s.io/utils/pointer"
+)
+
+type nodeOpt func(node *v1.Node)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/test_utils.go:29]: This is something you are adding in this refactoring, right?

Could you use this pattern instead? https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller_test.go#L2957

File: test/e2e/storage/csi_mock_volume.go [hunk 6]
```diff
@@ -1883,11 +1882,12 @@ const (
 	pvcAsSourceProtectionFinalizer = "snapshot.storage.kubernetes.io/pvc-as-source-protection"
 	volumeSnapshotContentFinalizer = "snapshot.storage.kubernetes.io/volumesnapshotcontent-bound-protection"
 	volumeSnapshotBoundFinalizer   = "snapshot.storage.kubernetes.io/volumesnapshot-bound-protection"
+	ErrReasonNotEnoughSpace        = "node(s) did not have enough free storage"
```
Comment (alculquicondor) [inline test/e2e/storage/csi_mock_volume.go:1885]: doesn't need to be exported

File: pkg/scheduler/framework/plugins/volumebinding/binder.go [hunk 7]
```diff
@@ -46,8 +46,8 @@ import (
 	"k8s.io/klog/v2"
 	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
 	pvutil "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util"
-	"k8s.io/kubernetes/pkg/controller/volume/scheduling/metrics"
 	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding/metrics"
 	volumeutil "k8s.io/kubernetes/pkg/volume/util"
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/binder.go:51]: Could you re-evaluate which types in this file need to be exported?

I suspect most of them don't need to be anymore. But you can leave that to a follow up PR.

Comment (ikeeip) [reply pkg/scheduler/framework/plugins/volumebinding/binder.go:51]: Let it be a follow up PR.  There are some inputs:
```
package volumebinding // import "k8s.io/kubernetes/pkg/scheduler/framework/plugins/volumebinding"

CONSTANTS

const ErrReasonBindConflict ConflictReason  - used in k8s.io/kubernetes packages
const ErrReasonNodeConflict ConflictReason  - used in k8s.io/kubernetes packages
const ErrReasonNotEnoughSpace string        - used only in current package (with this patch)
const ErrReasonPVNotExist string            - used only in current package
const DefaultBindTimeoutSeconds string      - Not used in k/k project
)

TYPES

type AssumeCache interface                      - used only in current package
func NewAssumeCache(...) AssumeCache            - used only in current package
type BindingInfo struct {}                      - used only in current package
type CapacityCheck struct {}                    - used only in current package
type ConflictReason string                      - used only in current package
type ConflictReasons []ConflictReason           - used in k8s.io/kubernetes packages
type FakeVolumeBinder struct {}                 - used only in current package
func NewFakeVolumeBinder(...) *FakeVolumeBinder - used in k8s.io/kubernetes packages
type FakeVolumeBinderConfig struct {}           - used in k8s.io/kubernetes packages
type PVAssumeCache interface {}                 - used only in current package
func NewPVAssumeCache(...) PVAssumeCache        - used only in current package
type PVCAssumeCache interface {}                - used only in current package
func NewPVCAssumeCache(...) PVCAssumeCache      - used only in current package
type PodVolumes struct {}                       - used only in current package
type SchedulerVolumeBinder interface {}         - used in k8s.io/kubernetes packages
func NewVolumeBinder(...) SchedulerVolumeBinder - used only in current package
type StorageResource struct {}                  - used only in current package
type VolumeBinding struct {}                    - used in k8s.io/kubernetes packages
type InTreeToCSITranslator interface {}         - used only in current package
```

File: test/e2e/storage/csi_mock_volume.go (hunk 6 above)
Comment (ikeeip) [reply test/e2e/storage/csi_mock_volume.go]: fixed

File: pkg/scheduler/framework/plugins/volumebinding/test_utils.go (hunk 5 above)
Comment (ikeeip) [reply pkg/scheduler/framework/plugins/volumebinding/test_utils.go]: Sure. Fixed.

File: pkg/scheduler/framework/plugins/volumebinding/test_utils.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	pb.Pod.Spec.Volumes = make([]v1.Volume, 0)
+	return pb
+}
+
+func (pb podBuilder) withNodeName(name string) podBuilder {
+	pb.Pod.Spec.NodeName = name
+	return pb
+}
+
+func (pb podBuilder) withNamespace(name string) podBuilder {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/test_utils.go:158]: I would just remove the prefix `with` to reduce line length. But this is a nit.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Comment by alculquicondor** [comment]
@msau42 PTAL

**Comment by msau42** [comment]
/lgtm
/approve

Thanks!

**Comment by Huang-Wei** [comment]
/hold

Would you mind squashing the commits?

**Comment by Huang-Wei** [comment]
@ikeeip it seems some code gets checked in and causes a conflict. Please rebase and resolve the conflicts.

**Comment by ikeeip** [comment]
> @ikeeip it seems some code gets checked in and causes a conflict. Please rebase and resolve the conflicts.

@Huang-Wei Done. Can you please /lgtm/approve again?

**Pull request merged**